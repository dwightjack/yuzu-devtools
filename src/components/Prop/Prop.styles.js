import { css } from 'emotion';

export const root = css`
  display: flex;
  align-items: flex-start;
  padding: 0 var(--gutter) calc(var(--gutter) / 2);
  font-family: var(--font-monospace);
  font-size: var(--font-size-s);
  font-weight: normal;
`;

export const label = css`
  display: inline-block;
  padding-right: 0.5em;
  color: var(--color-accent-secondary);
  line-height: 1.5;
  flex-shrink: 1;
  flex-grow: 0;
`;

export const labelFirst = css`
  padding-left: 1.4em;
`;

export const value = css`
  display: inline-block;
  font-family: var(--font-monospace);
  font-size: var(--font-size-s);
  font-weight: normal;
  line-height: 1.5;
  border: none;
  flex: 1 1 auto;
`;

export const stringStyle = css`
  &::after,
  &::before {
    content: '\\0022';
    display: inline-block;
  }
`;

export const inspect = css`
  padding: 0;
  border: 0;
  color: var(--color-quiet);
  visibility: hidden;

  [data-prop]:hover & {
    visibility: visible;
  }
`;
