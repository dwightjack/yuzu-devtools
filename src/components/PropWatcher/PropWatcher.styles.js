// import { css } from 'emotion';

// export const root = css`
//   position: relative;
//   cursor: pointer;
//   display: inline-block;
//   margin: -0.5em;
//   margin-right: -0.4em;
//   padding: 0.5em;
//   box-sizing: content-box;

//   & > svg {
//     visibility: hidden;
//     width: 1.25em;
//     height: 1.5em;
//     fill: var(--color-quiet);
//     vertical-align: middle;
//   }

//   &:hover > svg {
//     visibility: visible;
//   }

//   & > input:checked + svg {
//     visibility: visible;
//     fill: var(--color-accent-secondary);
//   }
// `;

// export const input = css`
//   position: absolute;
//   overflow: hidden;
//   width: 1px;
//   height: 1px;
//   padding: 0;
//   white-space: nowrap;
//   border: 0;
//   clip-path: inset(50%);
// `;
