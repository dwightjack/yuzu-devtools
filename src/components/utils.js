import isPlainObject from 'lodash/isPlainObject';

export const noop = () => {};

export const parseValue = (value) => {
  if (isPlainObject(value)) {
    return { type: 'object', value: '{...}', inspectable: true };
  }
  if (Array.isArray(value)) {
    return {
      type: 'array',
      value: `Array(${value.length})`,
      inspectable: true,
    };
  }
  if (typeof value === 'string') {
    if (value.startsWith('[function ')) {
      return { type: 'function', value };
    }
    return { type: 'string', value };
  }
  if (value === null || value === undefined) {
    return { type: 'other', value: String(value) };
  }
  return { type: 'other', value };
};
