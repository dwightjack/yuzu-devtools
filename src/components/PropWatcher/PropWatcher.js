import { html } from 'lit-html';
import { unsafeHTML } from 'lit-html/directives/unsafe-html';
import { component } from 'haunted';
import visibility from 'material-design-icons/action/svg/production/ic_visibility_24px.svg';
import { noop } from '../utils';
// import * as styles from './PropWatcher.styles';

export default function Watcher(props) {
  const { uid, key, watched = false, onToggle = noop } = props;

  return html`
    <style>
      :host {
        position: relative;
        display: inline-block;
        margin: -0.5em;
        margin-right: -0.4em;
        padding: 0.5em;
        box-sizing: content-box;
      }
      label {
        cursor: pointer;
      }
      svg {
        visibility: hidden;
        width: 1.25em;
        height: 1.5em;
        fill: var(--color-quiet);
        vertical-align: middle;
      }

      :hover > svg {
        visibility: visible;
      }

      input {
        position: absolute;
        overflow: hidden;
        width: 1px;
        height: 1px;
        padding: 0;
        white-space: nowrap;
        border: 0;
        clip-path: inset(50%);
      }

      input:checked + svg {
        visibility: visible;
        fill: var(--color-accent-secondary);
      }
    </style>
    <label title="Watch changes">
      <input
        type="checkbox"
        ?checked="${watched}"
        @click="${() => onToggle(uid, key, !watched)}"
        value="${`${uid}:${key}`}"
      />
      ${unsafeHTML(visibility)}
    </label>
  `;
}

Watcher.observedAttributes = ['watched', 'uid', 'key'];

customElements.define('yzdt-watcher', component(Watcher));
