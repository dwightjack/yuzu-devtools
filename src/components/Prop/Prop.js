import { html } from 'lit-html';
import { component } from 'haunted';
import { noop, parseValue } from '../utils';
import '../PropWatcher/PropWatcher';

export default function Prop({
  key,
  uid,
  value,
  watchable = false,
  watched = false,
  onWatchChange = noop,
}) {
  const { type, value: val /* , inspectable = false */ } = parseValue(value);
  return html`
    <style>
      :host {
        display: flex;
        align-items: flex-start;
        padding: 0 var(--gutter) calc(var(--gutter) / 2);
        font-family: var(--font-monospace);
        font-size: var(--font-size-s);
        font-weight: normal;
      }
      .label {
        display: inline-block;
        padding-right: 0.5em;
        color: var(--color-accent-secondary);
        line-height: 1.5;
        flex-shrink: 1;
        flex-grow: 0;
      }

      :host(:not([watchable])) .label {
        padding-left: 1.4em;
      }

      .value {
        display: inline-block;
        font-family: var(--font-monospace);
        font-size: var(--font-size-s);
        font-weight: normal;
        line-height: 1.5;
        border: none;
        flex: 1 1 auto;
      }

      .value[data-type='string']::after,
      .value[data-type='string']::before {
        content: '\\0022';
        display: inline-block;
      }
    </style>
    ${// prettier-ignore
    watchable
          ? html`
              <yzdt-watcher
                uid=${uid}
                key=${key}
                ?watched=${watched}
                @toggle=${({ detail }) => onWatchChange(uid, key, detail.watched)}
              ></yzdt-watcher>`
          : ''}<span
      class="label"
      >${key}: </span
    ><span class="value" data-type=${type}>${val}</span>
  `;
}

Prop.observedAttributes = ['key', 'uid', 'value', 'watched', 'watchable'];

customElements.define('yzdt-prop', component(Prop));
