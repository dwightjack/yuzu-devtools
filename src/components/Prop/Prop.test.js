import PropWatcher from '../PropWatcher/PropWatcher';
import * as utils from '../utils';

import Prop from './Prop';

jest.mock('../PropWatcher/PropWatcher', () => {
  const { mockComponent } = require('../test-utils');
  return jest.fn(mockComponent('PropWatcher'));
});
jest.mock('../utils', () => ({
  parseValue: jest.fn(),
  noop: () => {},
}));

describe('Prop', () => {
  let vals;
  const props = { key: 'name', uid: '_0', value: 'DEMO' };

  beforeAll(() => {
    vals = {
      type: 'string',
      value: 'DEMO',
    };
    utils.parseValue.mockImplementation(() => vals);
  });
  test('matches default snapshot', () => {
    expect(Prop({ ...props })).toMatchSnapshot();
  });
  test('calls "utils.parseValue"', () => {
    Prop({ ...props });
    expect(utils.parseValue).toHaveBeenCalledWith(props.value);
  });

  test('calls "PropWatcher" when watchable is true', () => {
    const onSelect = () => {};

    Prop({ ...props, onSelect, watchable: true, watched: true });
    Prop({ ...props, onSelect, watchable: false, watched: true });
    expect(PropWatcher).toHaveBeenCalledTimes(1);
    expect(PropWatcher).toHaveBeenCalledWith({
      key: props.key,
      uid: props.uid,
      watched: true,
      onSelect,
    });
  });
});
