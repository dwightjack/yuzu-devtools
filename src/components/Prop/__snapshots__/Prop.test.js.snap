// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prop matches default snapshot 1`] = `
<div
  class="css-1yfdty6"
  data-prop=""
  style=""
>
  
      
  
  
  
      
  <span
    class="css-68p123 css-1508h3c"
    style=""
  >
    name
    
    : 
  </span>
  <span
    class="css-fw3nso css-zeat2x"
    style=""
  >
    DEMO
    
  </span>
  
    
</div>
`;
